src/ui/access_dialog_ui.rs

```diff
@@ -0,0 +1,179 @@
+                && (*output != weak_output
+                    || buffer
+                        .as_ref()
+                        .map(|buffer| Scale::from(scale) == buffer.scale())
+                        .unwrap_or(false))
+        });
+
+        let Some(buffer) = buffers
+            .entry(weak_output)
+            .or_insert_with(|| render(scale).ok())
+        else {
+            buffers.clear();
+            return None;
+        };
+
+        let size = buffer.logical_size();
+        let buffer = TextureBuffer::from_memory_buffer(renderer.as_gles_renderer(), buffer).ok()?;
+
+        let location = (output_size.to_f64().to_point() - size.to_point()).downscale(2.);
+        let mut location = location.to_physical_precise_round(scale).to_logical(scale);
+        location.x = f64::max(0., location.x);
+        location.y = f64::max(0., location.y);
+
+        let elem = TextureRenderElement::from_texture_buffer(
+            buffer,
+            location,
+            1.,
+            None,
+            None,
+            Kind::Unspecified,
+        );
+        Some(PrimaryGpuTextureRenderElement(elem))
+    }
+
+    pub fn handle_key(&self, raw: Option<Keysym>, _mods: ModifiersState) {
+        if let Some(Keysym::Escape) = raw {
+            if let Some(request) = self.requests.borrow_mut().pop_front() {
+                let _ = request.deny();
+            }
+            self.buffers.borrow_mut().clear();
+        }
+
+        if let Some(Keysym::Return) = raw {
+            if let Some(request) = self.requests.borrow_mut().pop_front() {
+                let _ = request.accept();
+            }
+            self.buffers.borrow_mut().clear();
+        }
+    }
+}
+
+fn render(scale: f64) -> anyhow::Result<MemoryBuffer> {
+    let _span = tracy_client::span!("access_dialog_ui::render");
+
+    let padding: i32 = to_physical_precise_round(scale, PADDING);
+
+    let mut font = FontDescription::from_string(FONT);
+    font.set_absolute_size(to_physical_precise_round(scale, font.size()));
+
+    let surface = ImageSurface::create(cairo::Format::ARgb32, 0, 0)?;
+    let cr = cairo::Context::new(&surface)?;
+    let layout = pangocairo::functions::create_layout(&cr);
+    layout.context().set_round_glyph_positions(false);
+    layout.set_font_description(Some(&font));
+    layout.set_alignment(Alignment::Center);
+    layout.set_markup(TEXT);
+
+    let (mut width, mut height) = layout.pixel_size();
+    width += padding * 2;
+    height += padding * 2;
+
+    let surface = ImageSurface::create(cairo::Format::ARgb32, width, height)?;
+    let cr = cairo::Context::new(&surface)?;
+    cr.set_source_rgb(0.1, 0.1, 0.1);
+    cr.paint()?;
+
+    cr.move_to(padding.into(), padding.into());
+    let layout = pangocairo::functions::create_layout(&cr);
+    layout.context().set_round_glyph_positions(false);
+    layout.set_font_description(Some(&font));
+    layout.set_alignment(Alignment::Center);
+    layout.set_markup(TEXT);
```

Comment:
Guess it should use the title/subtitle/body from the request.

Reply:
Yeah, UI is completely unfinished, just something to get started discussing what is needed. There are also custom options which do not seem optional. Pointer/Touch input is also needed I guess.

Reply:
Okay, it now uses `title`,`subtitle` and the optional `body`. In addition it *should* also  use the options for the grant and deny labels. I left out the choices for now.